File: padlock-model/src/main/java/com/pyamsoft/padlock/model/sql/PadLockEntry.java
/*
 * Copyright 2016 Peter Kenji Yamanaka
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.pyamsoft.padlock.model.sql;

import android.database.sqlite.SQLiteOpenHelper;
import android.support.annotation.CheckResult;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import com.google.auto.value.AutoValue;
import com.squareup.sqldelight.RowMapper;
import com.squareup.sqldelight.SqlDelightStatement;

@AutoValue public abstract class PadLockEntry implements PadLockEntryModel {

  /**
   * The activity name of the PACKAGE entry in the database
   */
  @NonNull public static final String PACKAGE_ACTIVITY_NAME = "PACKAGE";
  @NonNull public static final String PACKAGE_EMPTY = "EMPTY";
  @NonNull public static final String ACTIVITY_EMPTY = "EMPTY";
  @SuppressWarnings("StaticInitializerReferencesSubClass") @NonNull public static final PadLockEntry
      EMPTY = new AutoValue_PadLockEntry(PACKAGE_EMPTY, ACTIVITY_EMPTY, null, 0, 0, false, false);
  @SuppressWarnings("StaticInitializerReferencesSubClass") @NonNull
  private static final Factory<PadLockEntry> FACTORY = new Factory<>(AutoValue_PadLockEntry::new);
  @NonNull public static final RowMapper<AllEntries> ALL_ENTRIES_MAPPER =
      FACTORY.all_entriesMapper(AutoValue_PadLockEntry_AllEntries::new);
  @NonNull public static final RowMapper<WithPackageName> WITH_PACKAGE_NAME_MAPPER =
      FACTORY.with_package_nameMapper(AutoValue_PadLockEntry_WithPackageName::new);
  @NonNull public static final Mapper<PadLockEntry> WITH_PACKAGE_ACTIVITY_NAME_DEFAULT_MAPPER =
      FACTORY.with_package_activity_name_defaultMapper();
  //@NonNull public static final RowMapper<WithPackageActivityName>
  //    WITH_PACKAGE_ACTIVITY_NAME_MAPPER =
  //    FACTORY.with_package_activity_nameMapper(AutoValue_PadLockEntry_WithPackageActivityName::new);

  //@CheckResult @NonNull
  //public static SqlDelightStatement withPackageActivityName(@NonNull String packageName,
  //    @NonNull String activityName) {
  //  return FACTORY.with_package_activity_name(packageName, activityName);
  //}

  @CheckResult @NonNull
  public static SqlDelightStatement withPackageActivityNameDefault(@NonNull String packageName,
      @NonNull String activityName) {
    return FACTORY.with_package_activity_name_default(packageName, PACKAGE_ACTIVITY_NAME,
        activityName, PACKAGE_ACTIVITY_NAME, activityName);
  }

  @CheckResult @NonNull
  public static SqlDelightStatement withPackageName(@NonNull String packageName) {
    return FACTORY.with_package_name(packageName);
  }

  @CheckResult @NonNull public static SqlDelightStatement queryAll() {
    return FACTORY.all_entries();
  }

  @CheckResult @NonNull
  public static PadLockEntry create(@NonNull String packageName, @NonNull String activityName,
      @Nullable String lockCode, long lockUntilTime, long ignoreUntilTime, boolean isSystem,
      boolean whitelist) {
    return FACTORY.creator.create(packageName, activityName, lockCode, lockUntilTime,
        ignoreUntilTime, isSystem, whitelist);
  }

  @CheckResult public static boolean isEmpty(@NonNull PadLockEntry entry) {
    return PACKAGE_EMPTY.equals(entry.packageName()) && ACTIVITY_EMPTY.equals(entry.activityName());
  }

  @CheckResult @NonNull
  public static InsertManager insertEntry(@NonNull SQLiteOpenHelper openHelper) {
    return new InsertManager(openHelper);
  }

  @CheckResult @NonNull
  public static DeletePackageManager deletePackage(@NonNull SQLiteOpenHelper openHelper) {
    return new DeletePackageManager(openHelper);
  }

  @CheckResult @NonNull public static DeletePackageActivityManager deletePackageActivity(
      @NonNull SQLiteOpenHelper openHelper) {
    return new DeletePackageActivityManager(openHelper);
  }

  @CheckResult @NonNull
  public static UpdateLockTimeManager updateLockTime(@NonNull SQLiteOpenHelper openHelper) {
    return new UpdateLockTimeManager(openHelper);
  }

  @CheckResult @NonNull
  public static UpdateIgnoreTimeManager updateIgnoreTime(@NonNull SQLiteOpenHelper openHelper) {
    return new UpdateIgnoreTimeManager(openHelper);
  }

  @CheckResult @NonNull
  public static UpdateWhitelistManager updateWhitelist(@NonNull SQLiteOpenHelper openHelper) {
    return new UpdateWhitelistManager(openHelper);
  }

  @AutoValue public static abstract class AllEntries implements All_entriesModel {

  }

  @AutoValue public static abstract class WithPackageName implements With_package_nameModel {

  }

  //@AutoValue public static abstract class WithPackageActivityName
  //    implements With_package_activity_nameModel {
  //
  //  @NonNull @CheckResult public static WithPackageActivityName empty() {
  //    return new AutoValue_PadLockEntry_WithPackageActivityName(PACKAGE_EMPTY, ACTIVITY_EMPTY, null,
  //        0, 0, false);
  //  }
  //
  //  @CheckResult public static boolean isEmpty(@NonNull WithPackageActivityName entry) {
  //    return PACKAGE_EMPTY.equals(entry.packageName()) && ACTIVITY_EMPTY.equals(
  //        entry.activityName());
  //  }
  //}

  @SuppressWarnings("WeakerAccess") public static class InsertManager {
    @NonNull private final Insert_entry insertEntry;

    InsertManager(@NonNull SQLiteOpenHelper openHelper) {
      this.insertEntry = new Insert_entry(openHelper.getWritableDatabase());
    }

    @CheckResult public long executeProgram(@NonNull PadLockEntry padLockEntry) {
      insertEntry.bind(padLockEntry.packageName(), padLockEntry.activityName(),
          padLockEntry.lockCode(), padLockEntry.lockUntilTime(), padLockEntry.ignoreUntilTime(),
          padLockEntry.systemApplication(), padLockEntry.whitelist());
      return insertEntry.program.executeInsert();
    }
  }

  @SuppressWarnings("WeakerAccess") public static class DeletePackageManager {
    @NonNull private final Delete_with_package_name deletePackage;

    DeletePackageManager(@NonNull SQLiteOpenHelper openHelper) {
      this.deletePackage = new Delete_with_package_name(openHelper.getWritableDatabase());
    }

    @CheckResult public int executeProgram(@NonNull String packageName) {
      deletePackage.bind(packageName);
      return deletePackage.program.executeUpdateDelete();
    }
  }

  @SuppressWarnings("WeakerAccess") public static class DeletePackageActivityManager {
    @NonNull private final Delete_with_package_activity_name deletePackageActivity;

    DeletePackageActivityManager(@NonNull SQLiteOpenHelper openHelper) {
      this.deletePackageActivity =
          new Delete_with_package_activity_name(openHelper.getWritableDatabase());
    }

    @CheckResult
    public int executeProgram(@NonNull String packageName, @NonNull String activityName) {
      deletePackageActivity.bind(packageName, activityName);
      return deletePackageActivity.program.executeUpdateDelete();
    }
  }

  @SuppressWarnings("WeakerAccess") public static class UpdateLockTimeManager {
    @NonNull private final Update_lock_until_time updateLockUntilTime;

    UpdateLockTimeManager(@NonNull SQLiteOpenHelper openHelper) {
      this.updateLockUntilTime = new Update_lock_until_time(openHelper.getWritableDatabase());
    }

    @CheckResult public int executeProgram(long time, @NonNull String packageName,
        @NonNull String activityName) {
      updateLockUntilTime.bind(time, packageName, activityName);
      return updateLockUntilTime.program.executeUpdateDelete();
    }
  }

  @SuppressWarnings("WeakerAccess") public static class UpdateIgnoreTimeManager {
    @NonNull private final Update_ignore_until_time updateIgnoreUntilTime;

    UpdateIgnoreTimeManager(@NonNull SQLiteOpenHelper openHelper) {
      this.updateIgnoreUntilTime = new Update_ignore_until_time(openHelper.getWritableDatabase());
    }

    @CheckResult public int executeProgram(long time, @NonNull String packageName,
        @NonNull String activityName) {
      updateIgnoreUntilTime.bind(time, packageName, activityName);
      return updateIgnoreUntilTime.program.executeUpdateDelete();
    }
  }

  @SuppressWarnings("WeakerAccess") public static class UpdateWhitelistManager {
    @NonNull private final Update_whitelist updateWhitelist;

    UpdateWhitelistManager(@NonNull SQLiteOpenHelper openHelper) {
      this.updateWhitelist = new Update_whitelist(openHelper.getWritableDatabase());
    }

    @CheckResult public int executeProgram(boolean whitelist, @NonNull String packageName,
        @NonNull String activityName) {
      updateWhitelist.bind(whitelist, packageName, activityName);
      return updateWhitelist.program.executeUpdateDelete();
    }
  }
}

File: padlock-base/src/main/java/com/pyamsoft/padlock/base/db/PadLockEntry.kt
/*
 *     Copyright (C) 2017 Peter Kenji Yamanaka
 *
 *     This program is free software; you can redistribute it and/or modify
 *     it under the terms of the GNU General Public License as published by
 *     the Free Software Foundation; either version 2 of the License, or
 *     (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *     GNU General Public License for more details.
 *
 *     You should have received a copy of the GNU General Public License along
 *     with this program; if not, write to the Free Software Foundation, Inc.,
 *     51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

package com.pyamsoft.padlock.base.db

import android.database.sqlite.SQLiteOpenHelper
import android.support.annotation.CheckResult
import com.google.auto.value.AutoValue
import com.pyamsoft.padlock.service.db.PadLockEntryModel
import com.squareup.sqldelight.RowMapper
import com.squareup.sqldelight.SqlDelightStatement

@AutoValue abstract class PadLockEntry : PadLockEntryModel {

  @AutoValue abstract class AllEntries : PadLockEntryModel.All_entriesModel

  @AutoValue abstract class WithPackageName : PadLockEntryModel.With_package_nameModel

  class InsertManager internal constructor(openHelper: SQLiteOpenHelper) {

    private val insertEntry = PadLockEntryModel.Insert_entry(openHelper.writableDatabase)

    @CheckResult
    fun executeProgram(padLockEntry: PadLockEntry): Long {
      insertEntry.bind(padLockEntry.packageName(), padLockEntry.activityName(),
          padLockEntry.lockCode(), padLockEntry.lockUntilTime(), padLockEntry.ignoreUntilTime(),
          padLockEntry.systemApplication(), padLockEntry.whitelist())
      return insertEntry.program.executeInsert()
    }
  }

  class DeletePackageManager internal constructor(openHelper: SQLiteOpenHelper) {

    private val deletePackage = PadLockEntryModel.Delete_with_package_name(
        openHelper.writableDatabase)

    @CheckResult
    fun executeProgram(packageName: String): Int {
      deletePackage.bind(packageName)
      return deletePackage.program.executeUpdateDelete()
    }
  }

  class DeletePackageActivityManager internal constructor(openHelper: SQLiteOpenHelper) {

    private val deletePackageActivity = PadLockEntryModel.Delete_with_package_activity_name(
        openHelper.writableDatabase)

    @CheckResult
    fun executeProgram(packageName: String, activityName: String): Int {
      deletePackageActivity.bind(packageName, activityName)
      return deletePackageActivity.program.executeUpdateDelete()
    }
  }

  class UpdateLockTimeManager internal constructor(openHelper: SQLiteOpenHelper) {

    private val updateLockUntilTime = PadLockEntryModel.Update_lock_until_time(
        openHelper.writableDatabase)

    @CheckResult
    fun executeProgram(time: Long, packageName: String,
        activityName: String): Int {
      updateLockUntilTime.bind(time, packageName, activityName)
      return updateLockUntilTime.program.executeUpdateDelete()
    }
  }

  class UpdateIgnoreTimeManager internal constructor(openHelper: SQLiteOpenHelper) {

    private val updateIgnoreUntilTime = PadLockEntryModel.Update_ignore_until_time(
        openHelper.writableDatabase)

    @CheckResult
    fun executeProgram(time: Long, packageName: String,
        activityName: String): Int {
      updateIgnoreUntilTime.bind(time, packageName, activityName)
      return updateIgnoreUntilTime.program.executeUpdateDelete()
    }
  }

  class UpdateWhitelistManager internal constructor(openHelper: SQLiteOpenHelper) {
    private val updateWhitelist = PadLockEntryModel.Update_whitelist(openHelper.writableDatabase)

    @CheckResult
    fun executeProgram(whitelist: Boolean, packageName: String,
        activityName: String): Int {
      updateWhitelist.bind(whitelist, packageName, activityName)
      return updateWhitelist.program.executeUpdateDelete()
    }
  }

  companion object {

    /**
     * The activity name of the PACKAGE entry in the database
     */
    const val PACKAGE_ACTIVITY_NAME = "PACKAGE"
    const internal val PACKAGE_EMPTY = "EMPTY"
    const internal val ACTIVITY_EMPTY = "EMPTY"

    private var insertManager: InsertManager? = null

    private var deletePackageManager: DeletePackageManager? = null

    private var deletePackageActivityManager: DeletePackageActivityManager? = null

    private var updateLockTimeManager: UpdateLockTimeManager? = null

    private var updateIgnoreTimeManager: UpdateIgnoreTimeManager? = null

    private var updateWhitelistManager: UpdateWhitelistManager? = null

    val EMPTY: PadLockEntry by lazy {
      AutoValue_PadLockEntry(PACKAGE_EMPTY, ACTIVITY_EMPTY, null, 0, 0,
          false, false)
    }

    private val FACTORY: PadLockEntryModel.Factory<PadLockEntry> by lazy<PadLockEntryModel.Factory<PadLockEntry>> {
      PadLockEntryModel.Factory { packageName, activityName, lockCode, lockUntilTime, ignoreUntilTime, systemApplication, whitelist ->
        AutoValue_PadLockEntry(packageName, activityName, lockCode, lockUntilTime, ignoreUntilTime,
            systemApplication, whitelist)
      }
    }

    internal val ALL_ENTRIES_MAPPER: RowMapper<AllEntries> by lazy<RowMapper<AllEntries>> {
      FACTORY.all_entriesMapper { packageName, activityName, whitelist ->
        AutoValue_PadLockEntry_AllEntries(packageName, activityName, whitelist)
      }
    }

    internal val WITH_PACKAGE_NAME_MAPPER: RowMapper<WithPackageName> by lazy<RowMapper<WithPackageName>> {
      FACTORY.with_package_nameMapper { activityName, whitelist ->
        AutoValue_PadLockEntry_WithPackageName(activityName, whitelist)
      }
    }

    internal val WITH_PACKAGE_ACTIVITY_NAME_DEFAULT_MAPPER: PadLockEntryModel.Mapper<PadLockEntry> by lazy {
      FACTORY.with_package_activity_name_defaultMapper()
    }

    @CheckResult
    internal fun withPackageActivityNameDefault(packageName: String,
        activityName: String): SqlDelightStatement {
      return FACTORY.with_package_activity_name_default(packageName, PACKAGE_ACTIVITY_NAME,
          activityName, PACKAGE_ACTIVITY_NAME, activityName)
    }

    @CheckResult internal fun withPackageName(packageName: String): SqlDelightStatement =
        FACTORY.with_package_name(packageName)

    @CheckResult internal fun queryAll(): SqlDelightStatement = FACTORY.all_entries()

    @CheckResult
    fun create(packageName: String, activityName: String,
        lockCode: String?, lockUntilTime: Long, ignoreUntilTime: Long, isSystem: Boolean,
        whitelist: Boolean): PadLockEntry {
      return FACTORY.creator.create(packageName, activityName, lockCode, lockUntilTime,
          ignoreUntilTime, isSystem, whitelist)
    }

    @CheckResult
    fun isEmpty(entry: PadLockEntry): Boolean =
        PACKAGE_EMPTY == entry.packageName() && ACTIVITY_EMPTY == entry.activityName()

    @CheckResult internal fun insertEntry(openHelper: SQLiteOpenHelper): InsertManager {
      val obj: InsertManager? = insertManager
      return if (obj == null) {
        val im = InsertManager(openHelper)
        insertManager = im

        // Return
        im
      } else {
        // Return
        obj
      }
    }

    @CheckResult
    internal fun deletePackage(openHelper: SQLiteOpenHelper): DeletePackageManager {
      val obj: DeletePackageManager? = deletePackageManager
      return if (obj == null) {
        val dpm = DeletePackageManager(openHelper)
        deletePackageManager = dpm

        // Return
        dpm
      } else {
        // Return
        obj
      }
    }

    @CheckResult
    internal fun deletePackageActivity(
        openHelper: SQLiteOpenHelper): DeletePackageActivityManager {
      val obj: DeletePackageActivityManager? = deletePackageActivityManager
      return if (obj == null) {
        val dpam = DeletePackageActivityManager(openHelper)
        deletePackageActivityManager = dpam

        // Return
        dpam
      } else {
        // Return
        obj
      }
    }

    @CheckResult
    internal fun updateLockTime(openHelper: SQLiteOpenHelper): UpdateLockTimeManager {
      val obj: UpdateLockTimeManager? = updateLockTimeManager
      return if (obj == null) {
        val ultm = UpdateLockTimeManager(openHelper)
        updateLockTimeManager = ultm

        // Return
        ultm
      } else {
        // Return
        obj
      }
    }

    @CheckResult
    internal fun updateIgnoreTime(openHelper: SQLiteOpenHelper): UpdateIgnoreTimeManager {
      val obj: UpdateIgnoreTimeManager? = updateIgnoreTimeManager
      return if (obj == null) {
        val uitm = UpdateIgnoreTimeManager(openHelper)
        updateIgnoreTimeManager = uitm
        // Return
        uitm
      } else {
        // Return
        obj
      }
    }

    @CheckResult
    internal fun updateWhitelist(openHelper: SQLiteOpenHelper): UpdateWhitelistManager {
      val obj: UpdateWhitelistManager? = updateWhitelistManager
      return if (obj == null) {
        val uwm = UpdateWhitelistManager(openHelper)
        // Return
        updateWhitelistManager = uwm
        uwm
      } else {
        // Return
        obj
      }
    }

    internal fun reset() {
      deletePackageManager = null
      deletePackageActivityManager = null
      insertManager = null
      updateLockTimeManager = null
      updateWhitelistManager = null
      updateIgnoreTimeManager = null
    }
  }
}
